File: client/typescript/gamium-client-common/src/condition.ts
import { GamiumClient } from './gamium-client';
import { UIElement } from './object/ui-element';
import { ObjectInfo } from './protocols/types';

type ConditionFn<Type> = (driver: GamiumClient) => PromiseLike<Type>;
export class Condition<Type> {
  constructor(public readonly message: string, public readonly contidionFunc: ConditionFn<Type>) {}
}

export class ObjectInfoCondition extends Condition<ObjectInfo> {
  private _nominal: undefined;
}

export class ObjectInfosCondition extends Condition<ObjectInfo[]> {
  private _nominal: undefined;
}

export class UIElementCondition extends Condition<UIElement> {
  private _nominal: undefined;
}

File: client/python/gamium/gamium/condition/condition.py
from typing import Any, Callable, Generic, List
from gamium.protocol.types import ObjectInfo
from gamium.utils.generics import T

ConditionFn = Callable[[Any], T]  # Callable[[IGamiumClient], T]


class Condition(Generic[T]):
    def __init__(self, message: str, func: ConditionFn[T]):
        self.message = message
        self.func = func


class ObjectInfoCondition(Condition[ObjectInfo]):
    pass


class ObjectInfosCondition(Condition[List[ObjectInfo]]):
    pass
